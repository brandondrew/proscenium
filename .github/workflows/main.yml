name: Ruby

on:
  push:
    branches:
      - master

  pull_request:

jobs:
  release:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v3

      - name: Set up Ruby
        uses: ruby/setup-ruby@v1
        with:
          ruby-version: 3.1

      - uses: denoland/setup-deno@v1.0.0
        with:
          deno-version: v1.x

      - name: Compile esbuild
        run: deno task compile:esbuild:macm

      - name: Install gem dependencies
        run: bundle install

      - name: Build native gem
        env:
          PROFILE: release
        run: bundle exec rake build

      - uses: actions/upload-artifact@v3
        with:
          name: artifacts
          path: bin/esbuild-*
