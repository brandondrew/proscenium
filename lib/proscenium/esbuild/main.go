package main

// go build -buildmode=c-shared -o main.so main.go

import "C"
import "os"
import "strings"
import "encoding/json"

// import "fmt"
import "github.com/evanw/esbuild/pkg/api"

//export transform
func transform(input *C.char) *C.char {
	inputString := C.GoString(input)
	result := api.Transform(inputString, api.TransformOptions{
		Loader: api.LoaderJS,
	})

	s := string(result.Code)

	return C.CString(s)
}

var envPlugin = api.Plugin{
	Name: "env",
	Setup: func(build api.PluginBuild) {
		// Intercept import paths called "env" so esbuild doesn't attempt
		// to map them to a file system location. Tag them with the "env-ns"
		// namespace to reserve them for this plugin.
		build.OnResolve(api.OnResolveOptions{Filter: `^env$`},
			func(args api.OnResolveArgs) (api.OnResolveResult, error) {
				return api.OnResolveResult{
					Path:      args.Path,
					Namespace: "env-ns",
				}, nil
			})

		// Load paths tagged with the "env-ns" namespace and behave as if
		// they point to a JSON file containing the environment variables.
		build.OnLoad(api.OnLoadOptions{Filter: `.*`, Namespace: "env-ns"},
			func(args api.OnLoadArgs) (api.OnLoadResult, error) {
				mappings := make(map[string]string)
				for _, item := range os.Environ() {
					if equals := strings.IndexByte(item, '='); equals != -1 {
						mappings[item[:equals]] = item[equals+1:]
					}
				}
				bytes, err := json.Marshal(mappings)
				if err != nil {
					return api.OnLoadResult{}, err
				}
				contents := string(bytes)
				return api.OnLoadResult{
					Contents: &contents,
					Loader:   api.LoaderJSON,
				}, nil
			})
	},
}

//export build
func build(entrypoint *C.char) *C.char {
	entrypointString := C.GoString(entrypoint)

	result := api.Build(api.BuildOptions{
		EntryPoints: []string{entrypointString},
		Bundle:      true,
		Plugins:     []api.Plugin{envPlugin},
		Write:       false,
	})

	outputString := string(result.OutputFiles[0].Contents)
	return C.CString(outputString)
}

func main() {
	// fmt.Println(transform())
	// fmt.Printf("%s", build("input.ts"))
}
