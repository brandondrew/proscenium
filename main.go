package main

/*
struct Result {
	int success;
	char* response;
};
*/
import "C"

import (
	"encoding/json"
	"joelmoss/proscenium/internal/builder"
	"joelmoss/proscenium/internal/resolver"
	"joelmoss/proscenium/internal/types"
	"path"
	"strings"
)

// Build the given `path` in the `root`.
//
//   - path - The path to build relative to `root`.
//   - root - The working directory.
//   - env - The environment (1 = development, 2 = test, 3 = production)
//   - importMap - Path to the import map relative to `root`.
//   - debug
//
//export build
func build(filepath *C.char, root *C.char, env C.uint, importMap *C.char, debug bool) C.struct_Result {
	types.Env = types.Environment(env)

	pathStr := C.GoString(filepath)

	result := builder.Build(builder.BuildOptions{
		Path:          pathStr,
		Root:          C.GoString(root),
		ImportMapPath: C.GoString(importMap),
		Debug:         debug,
	})

	if len(result.Errors) != 0 {
		j, err := json.Marshal(result.Errors[0])
		if err != nil {
			return C.struct_Result{C.int(0), C.CString(string(err.Error()))}
		}

		return C.struct_Result{C.int(0), C.CString(string(j))}
	}

	contents := string(result.OutputFiles[0].Contents)

	isSourceMap := strings.HasSuffix(pathStr, ".map")
	if isSourceMap {
		return C.struct_Result{C.int(1), C.CString(contents)}
	}

	sourcemapUrl := path.Base(pathStr)
	if strings.HasSuffix(result.OutputFiles[0].Path, ".css") {
		contents += "/*# sourceMappingURL=" + sourcemapUrl + ".map */"
	} else {
		contents += "//# sourceMappingURL=" + sourcemapUrl + ".map"
	}

	return C.struct_Result{C.int(1), C.CString(contents)}
}

// Resolve the given `path` relative to the `root`.
//
//   - path - The path to build relative to `root`.
//   - root - The working directory.
//   - env - The environment (1 = development, 2 = test, 3 = production)
//   - importMap - Path to the import map relative to `root`.
//
//export resolve
func resolve(path *C.char, root *C.char, env C.uint, importMap *C.char) C.struct_Result {
	types.Env = types.Environment(env)

	resolvedPath, err := resolver.Resolve(resolver.Options{
		Path:          C.GoString(path),
		Root:          C.GoString(root),
		ImportMapPath: C.GoString(importMap),
	})
	if err != nil {
		return C.struct_Result{C.int(0), C.CString(string(err.Error()))}
	}

	return C.struct_Result{C.int(1), C.CString(resolvedPath)}
}

func main() {}
