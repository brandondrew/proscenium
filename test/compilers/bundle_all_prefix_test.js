import { assertStringIncludes, AssertionError } from 'testing/asserts.ts'
import { assertSnapshot } from 'testing/snapshot.ts'
import { beforeEach, describe, it } from 'testing/bdd.ts'
import { join } from 'std/path/mod.ts'

import main from '../../lib/proscenium/compilers/esbuild.js'

const root = join(Deno.cwd(), 'test', 'internal')
const lightningcssBin = join(Deno.cwd(), 'bin', 'lightningcss')

function assertStringExcludes(actual, expected, msg) {
  if (actual.includes(expected)) {
    if (!msg) {
      msg = `actual: "${actual}" expected NOT to contain: "${expected}"`
    }
    throw new AssertionError(msg)
  }
}

describe('bundle-all: prefix', () => {
  beforeEach(() => {
    Deno.env.set('RAILS_ENV', 'test')
    Deno.env.set('PROSCENIUM_TEST', 'test')
  })

  it('js import', async t => {
    const result = await main('lib/bundle_all_import/index.js', {
      root,
      lightningcssBin
    })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('css', async t => {
    const result = await main('lib/bundle_all_import/index.css', {
      root,
      lightningcssBin
    })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('tree shaking', async () => {
    const result = await main('lib/bundle_all_import/tree_shaking.js', {
      root,
      lightningcssBin
    })

    const code = new TextDecoder().decode(result)

    assertStringIncludes(code, 'console.log("foo")')
    assertStringIncludes(code, 'console.log("foo2")')
    assertStringExcludes(code, 'console.log("bar")')
  })

  it('with import map', async () => {
    const result = await main('lib/bundle_all_import/import_map.js', {
      root,
      lightningcssBin,
      importMap: 'config/import_maps/simple.json'
    })

    const code = new TextDecoder().decode(result)

    assertStringIncludes(code, 'console.log("/lib/foo.js");')
  })
})
