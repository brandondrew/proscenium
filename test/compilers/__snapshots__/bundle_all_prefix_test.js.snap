export const snapshot = {};

snapshot[`bundle-all: prefix > js import 1`] = `
'// bundleAll:/Users/joelmoss/dev/proscenium/test/internal/lib/bundle_all_import/three.js
var three_default = () => console.log("three");

// bundleAll:/Users/joelmoss/dev/proscenium/test/internal/lib/bundle_all_import/two.js
var two_default = () => {
  console.log("two");
  three_default();
};

// url:https://esm.sh/v99/ip-regex@5.0.0/es2022/ip-regex.js
var g = "[a-fA-F\\\\\\\\d:]";
var d = (e) => e && e.includeBoundaries ? \`(?:(?<=\\\\\\\\s|^)(?=\${g})|(?<=\${g})(?=\\\\\\\\s|\$))\` : "";
var c = "(?:25[0-5]|2[0-4]\\\\\\\\d|1\\\\\\\\d\\\\\\\\d|[1-9]\\\\\\\\d|\\\\\\\\d)(?:\\\\\\\\.(?:25[0-5]|2[0-4]\\\\\\\\d|1\\\\\\\\d\\\\\\\\d|[1-9]\\\\\\\\d|\\\\\\\\d)){3}";
var \$ = "[a-fA-F\\\\\\\\d]{1,4}";
var x = \`
(?:
(?:\${\$}:){7}(?:\${\$}|:)|                                    // 1:2:3:4:5:6:7::  1:2:3:4:5:6:7:8
(?:\${\$}:){6}(?:\${c}|:\${\$}|:)|                             // 1:2:3:4:5:6::    1:2:3:4:5:6::8   1:2:3:4:5:6::8  1:2:3:4:5:6::1.2.3.4
(?:\${\$}:){5}(?::\${c}|(?::\${\$}){1,2}|:)|                   // 1:2:3:4:5::      1:2:3:4:5::7:8   1:2:3:4:5::8    1:2:3:4:5::7:1.2.3.4
(?:\${\$}:){4}(?:(?::\${\$}){0,1}:\${c}|(?::\${\$}){1,3}|:)| // 1:2:3:4::        1:2:3:4::6:7:8   1:2:3:4::8      1:2:3:4::6:7:1.2.3.4
(?:\${\$}:){3}(?:(?::\${\$}){0,2}:\${c}|(?::\${\$}){1,4}|:)| // 1:2:3::          1:2:3::5:6:7:8   1:2:3::8        1:2:3::5:6:7:1.2.3.4
(?:\${\$}:){2}(?:(?::\${\$}){0,3}:\${c}|(?::\${\$}){1,5}|:)| // 1:2::            1:2::4:5:6:7:8   1:2::8          1:2::4:5:6:7:1.2.3.4
(?:\${\$}:){1}(?:(?::\${\$}){0,4}:\${c}|(?::\${\$}){1,6}|:)| // 1::              1::3:4:5:6:7:8   1::8            1::3:4:5:6:7:1.2.3.4
(?::(?:(?::\${\$}){0,5}:\${c}|(?::\${\$}){1,7}|:))             // ::2:3:4:5:6:7:8  ::2:3:4:5:6:7:8  ::8             ::1.2.3.4
)(?:%[0-9a-zA-Z]{1,})?                                             // %eth0            %1
\`.replace(/\\\\s*\\\\/\\\\/.*\$/gm, "").replace(/\\
/g, "").trim();
var n = new RegExp(\`(?:^\${c}\$)|(?:^\${x}\$)\`);
var t = new RegExp(\`^\${c}\$\`);
var v = new RegExp(\`^\${x}\$\`);
var a = (e) => e && e.exact ? n : new RegExp(\`(?:\${d(e)}\${c}\${d(e)})|(?:\${d(e)}\${x}\${d(e)})\`, "g");
a.v4 = (e) => e && e.exact ? t : new RegExp(\`\${d(e)}\${c}\${d(e)}\`, "g");
a.v6 = (e) => e && e.exact ? v : new RegExp(\`\${d(e)}\${x}\${d(e)}\`, "g");
var r = a;

// bundleAll:/Users/joelmoss/dev/proscenium/test/internal/lib/bundle_all_import/one.js
var one_default = () => {
  console.log("one");
  two_default();
  r();
};

// lib/bundle_all_import/index.js
import foo4 from "/lib/foo4.js";
one_default();
foo4();
//# sourceMappingURL=index.js.map
'
`;

snapshot[`bundle-all: prefix > css 1`] = `
"/* url:https://unpkg.com/open-props@1.5.3/blue.min.css */
:where(html) {
  --blue-0:#e7f5ff;
  --blue-1:#d0ebff;
  --blue-2:#a5d8ff;
  --blue-3:#74c0fc;
  --blue-4:#4dabf7;
  --blue-5:#339af0;
  --blue-6:#228be6;
  --blue-7:#1c7ed6;
  --blue-8:#1971c2;
  --blue-9:#1864ab;
  --blue-10:#145591;
  --blue-11:#114678;
  --blue-12:#0d375e;
}

/* url:https://unpkg.com/open-props@1.5.3/green.min.css */
:where(html) {
  --green-0:#ebfbee;
  --green-1:#d3f9d8;
  --green-2:#b2f2bb;
  --green-3:#8ce99a;
  --green-4:#69db7c;
  --green-5:#51cf66;
  --green-6:#40c057;
  --green-7:#37b24d;
  --green-8:#2f9e44;
  --green-9:#2b8a3e;
  --green-10:#237032;
  --green-11:#1b5727;
  --green-12:#133d1b;
}

/* url:https://unpkg.com/open-props@1.5.3/red.min.css */
:where(html) {
  --red-0:#fff5f5;
  --red-1:#ffe3e3;
  --red-2:#ffc9c9;
  --red-3:#ffa8a8;
  --red-4:#ff8787;
  --red-5:#ff6b6b;
  --red-6:#fa5252;
  --red-7:#f03e3e;
  --red-8:#e03131;
  --red-9:#c92a2a;
  --red-10:#b02525;
  --red-11:#962020;
  --red-12:#7d1a1a;
}

/* bundleAll:/Users/joelmoss/dev/proscenium/test/internal/lib/bundle_all_import/two.css */
body {
  color: green;
}

/* bundleAll:/Users/joelmoss/dev/proscenium/test/internal/lib/bundle_all_import/one.css */

/* lib/bundle_all_import/index.css */
body {
  color: red;
}
/*# sourceMappingURL=index.css.map */
"
`;
